src/Qir/Runtime/lib/Simulators/ToffoliSimulator.cpp

```diff
@@ -109,11 +109,22 @@ namespace Quantum
             return std::abs(actualZeroProbability - probabilityOfZero) < precision;
         }
 
+        // To do: remove after the `function DumpMachine<'T> (location : 'T) : Unit` is implemented.
         void GetState(TGetStateCallback callback) override
         {
             throw std::logic_error("operation_not_supported");
         }
 
+        void DumpMachine(const void* location) override
+        {
+            throw std::logic_error("operation_not_supported");

```

Comment:
Is there a reason this isn't implemented? See https://github.com/microsoft/qsharp-runtime/blob/21f5fab97682d9bc0ae8e3b58f679fdb952f1c24/src/Simulation/Simulators/ToffoliSimulator/ToffoliSimulator.cs#L336 for an example of what diagnostic output for Toffoli simulation should look like.

In any case, this shouldn't throw an exception, but rather should provide a clear message at the console informing the user that this diagnostic is not yet supported.

Reply:
Thank you for the direct link. But I don't know C# and it is very hard for me to translate the C# code to C++. May be you can provide a couple of examples of the eventual output? Or can refer me to how to redirect the Q# project to use a Toffoli simulator (with C# runtime)? (such that I can generate the example output by myself)

I have written these fragments (exception throw) by example of `ToffoliSimulator::GetState()` (one function above), and based on observation that currently in QIR the `ToffoliSimulator` is only used in tests. I thought the `ToffoliSimulator` is something dying (or already dead), but looks like it is not.

Reply:
The Toffoli simulator is rather important to a number of different domain areas, especially reversible oracle synthesis. As such, I would suggest that it's important to make sure that even if a feature isn't implemented, the Toffoli simulator at least fails gracefully — in this case, replacing the `throw` with an insertion to `std::cout` may be enough to do that.

As for implementing the feature, the Toffoli simulator takes advantage of the fact that if you start in a computational basis state and only apply `X` and `Controlled X` operations, the state of the system always has the form |𝑏₀𝑏₁…𝑏ₙ₋₁⟩ for some classical bitstring 𝑏₀𝑏₁…𝑏ₙ₋₁. Thus, a reasonable diagnostic for Toffoli simulation is to display the state 𝑏₀𝑏₁…𝑏ₙ₋₁. When 𝑛 is large enough, then hex output can be particularly helpful; most of the C# code above is intended to switch between outputting binary and hex formats based on the size of the register.

Reply:
For now changing to
```c++
        void DumpMachine(const void* location) override
        {
            std::cerr << __func__ << " is not yet implemented" << std::endl;
            // TODO: How to implement: https://github.com/microsoft/qsharp-runtime/pull/634#discussion_r615183791
        }

        void DumpRegister(const void* location, const QirArray* qubits) override
        {
            std::cerr << __func__ << " is not yet implemented" << std::endl;
            // TODO: How to implement: https://github.com/microsoft/qsharp-runtime/pull/634#discussion_r615183791
        }
```